trpcultivate_phenotypes/src/TripalCultivateValidator/ValidatorTraits/Headers.php

```diff
@@ -0,0 +1,213 @@
+    }
+    else {
+      throw new \Exception('Cannot retrieve ' . $type_key . ' headers from the context array as one has not been set by setHeaders() method.');
+    }
+  }
+  
+  /**
+   * Get optional headers.
+   * 
+   * @return array
+   *   All headers of type optional, keyed by the index (order) from
+   *   the headers array and header name as the value.
+   * 
+   *   The key optional in the context headers array set by the setter method.
+   *   NOTE: the headers array is zero-based index.
+   * 
+   * @throws \Exception
+   *  - If the 'headers' key does not exists in the context array
+   *    (ie. the headers element has NOT been set).
+   */
+  public function getOptionalHeaders() {
+    
+    $type_key = 'optional';
+
+    if (array_key_exists($this->context_key, $this->context)) {
+      return $this->context[ $this->context_key ][ $type_key ];
+    }
+    else {
+      throw new \Exception('Cannot retrieve ' . $type_key . ' headers from the context array as one has not been set by setHeaders() method.');
+    }
+  }
+
+  /**
+   * Get headers of type(s).
+   * 
+   * @param array $types
+   *   A list of header types to get. 
+   *   Default to required and optional header types.
+   * 
+   * @return array
+   *   A single list of headers matching a type defined by the types parameter.
+   *   For example, if the types parameter includes required and optional the the 
+   *   resulting array will container all headers in order of index assuming there
+   *   are only required and optional types supported.
+   *
+   *   The returned array is keyed by the index (column order) from
+   *   the headers array and header name as the value.
+   *   NOTE: the headers array is zero-based index.
+   * 
+   * @throws \Exception
+   *  - If the 'headers' key does not exists in the context array
+   *    (ie. the headers element has NOT been set).
+   *  - If an unrecognized header type is requested in the types parameter.
+   */
+  public function getHeaders(array $types = ['required', 'optional']) {
+
+    $valid_types = $this->types;
+    
+    // Pull out any unrecognized types by compairing the types paramter
+    // to the types property listing valid types.
+    $invalid_types = array_filter($types, function($type) use($valid_types) { 
+      return !in_array($type, $valid_types); 
+    });
+    
+    // If any such unrecognized types are detected, throw an exception.
+    if (!empty($invalid_types)) {
+      $str_invalid_types = implode(', ', $invalid_types);
+      $str_valid_types = implode(', ', $valid_types);
+      throw new \Exception('Cannot retrieve invalid header types: ' . $str_invalid_types . '. Use one of valid types: [' . $str_valid_types . ']');
+    }
+
+    if (array_key_exists($this->context_key, $this->context)) {
+      // At this point, types requested are valid.   
+      $headers = [];
+
+      foreach($types as $type) {
+        foreach($this->context[ $this->context_key ][ $type ] as $index => $header) {
```

Comment:
Similarly to getRequiredHeaders() and getOptionalHeaders(), we want to check for the presence of each 'type' key in the `$this->context['headers']` array